packages/opentelemetry-exporter-collector/src/CollectorExporter.ts

```diff
@@ -0,0 +1,139 @@
+
+import * as collectorTypes from './types';
+import { toCollectorSpan } from './transform';
+
+import { onInit, onShutdown, sendSpans } from './platform/index';
+
+/**
+ * Collector Exporter Config
+ */
+export interface CollectorExporterConfig {
+  hostName?: string;
+  logger?: Logger;
+  serviceName?: string;
+  spanAttributes?: Attributes;
+  url?: string;
+}
+
+const defaultServiceName = 'collector-exporter';
+const defaultCollectorUrl = 'http://localhost:55678/v1/trace';
+
+/**
+ * Collector Exporter
+ */
+export class CollectorExporter implements SpanExporter {
+  readonly serviceName: string;
+  readonly url: string;
+  readonly logger: Logger;
+  readonly hostName: string | undefined;
+  private _isShutdown: boolean = false;
+
+  /**
+   * @param config
+   */
+  constructor(config: CollectorExporterConfig = {}) {
+    this.serviceName = config.serviceName || defaultServiceName;
+    this.url = config.url || defaultCollectorUrl;
+    if (typeof config.hostName === 'string') {
+      this.hostName = config.hostName;
+    }
+
+    this.logger = config.logger || new NoopLogger();
+
+    this.shutdown = this.shutdown.bind(this);
+
+    // platform dependent
+    onInit(this.shutdown);
+  }
+
+  /**
+   * Export spans.
+   * @param spans
+   * @param resultCallback
+   */
+  export(
+    spans: ReadableSpan[],
+    resultCallback: (result: ExportResult) => void
+  ) {
+    this._exportSpans(spans)
+      .then(() => {
+        resultCallback(ExportResult.SUCCESS);
+      })
+      .catch((status: number = 0) => {
+        if (status < 500) {
+          resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+        } else {
+          resultCallback(ExportResult.FAILED_RETRYABLE);
+        }
+      });
+  }
+
+  private _exportSpans(spans: ReadableSpan[]): Promise<unknown> {
+    return new Promise((resolve, reject) => {
+      try {
+        const spansToBeSent: collectorTypes.Span[] = spans.map(span =>
+          toCollectorSpan(span)
+        );
+        this.logger.debug('spans to be sent', spansToBeSent);
+        this.sendSpan(spansToBeSent, resolve, reject);
+      } catch (e) {
+        reject(e);
+      }
+    });
+  }
+
+  /**
+   * Send spans to [opentelemetry collector]{@link https://github.com/open-telemetry/opentelemetry-collector}
+   *     it will use the appropriate transport layer automatically depends on platform
+   * @param spans
+   * @param onSuccess
+   * @param onError
+   */
+  sendSpan(
+    spans: collectorTypes.Span[],
+    onSuccess: () => void,
+    onError: (status?: number) => void
+  ) {
+    // platform dependent
+    sendSpans(spans, onSuccess, onError, this);
+  }
+
+  /**
+   * Shutdown the exporter.
+   */
+  shutdown(): void {
+    if (this._isShutdown) {
+      this.logger.debug('shutdown already started');
+      return;
+    }
+    this._isShutdown = true;
+    this.logger.debug('shutdown started');
+
+    // platform dependent
+    onShutdown(this.shutdown);
+
+    this._exportSpans([]).then(() => {
```

Comment:
I think this is added to do final optimistic export, but the list is empty. Do you plan to add support for this later?

Reply:
Additionally, if this is in response to an `unload` event, do we want to make the XHR call synchronously? I know it tries to do `navigator.sendBeacon` first, but if it falls back to XHR in an unload event handler, async XHR requests may not finish.

Reply:
I mean so far nothing yet will happen as you mentioned, I added @TODO to support that for batch span, the same as it is for zipkin, however now I'm not sure how yet. I guess this can be done together with zipkin then in separate PR. just wanted to keep consistency for now.